File: frontend/src/lib/api/runtime.ts
import { type IdParam, type RequestQueryValue, requestV2, toPathSegment } from "../api-client";
import type {
  RuntimeApprovalActionInput,
  RuntimeApprovalActionRead,
  RuntimeApprovalListParams,
  RuntimeApprovalListRead,
  RuntimeApprovalRead,
  RuntimeArtifactRead,
  RuntimeCancelRead,
  RuntimeRunCreateInput,
  RuntimeRunCreated,
  RuntimeRunListParams,
  RuntimeRunListRead,
  RuntimeRunRead,
  RuntimeTraceEventListParams,
  RuntimeTraceEventListRead,
} from "../types/runtime";

function runtimeRunPath(runId: IdParam): string {
  return `/runtime/runs/${toPathSegment(runId)}`;
}

function runtimeApprovalPath(approvalId: IdParam): string {
  return `/runtime/approvals/${toPathSegment(approvalId)}`;
}

function toQueryRecord<T extends object>(
  params?: T,
): Record<string, RequestQueryValue> | undefined {
  return params as Record<string, RequestQueryValue> | undefined;
}

export function listRuntimeRuns(
  params?: RuntimeRunListParams,
  signal?: AbortSignal,
): Promise<RuntimeRunListRead> {
  return requestV2<RuntimeRunListRead>("/runtime/runs", {
    query: toQueryRecord(params),
    signal,
  });
}

export function createRuntimeRun(
  payload: RuntimeRunCreateInput,
  signal?: AbortSignal,
): Promise<RuntimeRunCreated> {
  return requestV2<RuntimeRunCreated>("/runtime/runs", {
    body: payload,
    method: "POST",
    signal,
  });
}

export function getRuntimeRun(runId: IdParam, signal?: AbortSignal): Promise<RuntimeRunRead> {
  return requestV2<RuntimeRunRead>(runtimeRunPath(runId), { signal });
}

export function getRuntimeRunArtifact(
  runId: IdParam,
  signal?: AbortSignal,
): Promise<RuntimeArtifactRead> {
  return requestV2<RuntimeArtifactRead>(`${runtimeRunPath(runId)}/artifacts`, { signal });
}

export function getRuntimeRunTrace(
  runId: IdParam,
  signal?: AbortSignal,
): Promise<RuntimeTraceEventListRead> {
  return requestV2<RuntimeTraceEventListRead>(`${runtimeRunPath(runId)}/trace`, { signal });
}

export function cancelRuntimeRun(
  runId: IdParam,
  signal?: AbortSignal,
): Promise<RuntimeCancelRead> {
  return requestV2<RuntimeCancelRead>(`${runtimeRunPath(runId)}/cancel`, {
    method: "POST",
    signal,
  });
}

export function listRuntimeApprovals(
  params?: RuntimeApprovalListParams,
  signal?: AbortSignal,
): Promise<RuntimeApprovalListRead> {
  return requestV2<RuntimeApprovalListRead>("/runtime/approvals", {
    query: toQueryRecord(params),
    signal,
  });
}

export function getRuntimeApproval(
  approvalId: IdParam,
  signal?: AbortSignal,
): Promise<RuntimeApprovalRead> {
  return requestV2<RuntimeApprovalRead>(runtimeApprovalPath(approvalId), { signal });
}

export function listRuntimeTraceEvents(
  params?: RuntimeTraceEventListParams,
  signal?: AbortSignal,
): Promise<RuntimeTraceEventListRead> {
  return requestV2<RuntimeTraceEventListRead>("/runtime/trace-events", {
    query: toQueryRecord(params),
    signal,
  });
}

export function approveRuntimeApproval(
  approvalId: IdParam,
  payload: RuntimeApprovalActionInput,
  signal?: AbortSignal,
): Promise<RuntimeApprovalActionRead> {
  return requestV2<RuntimeApprovalActionRead>(`${runtimeApprovalPath(approvalId)}/approve`, {
    body: payload,
    method: "POST",
    signal,
  });
}

export function denyRuntimeApproval(
  approvalId: IdParam,
  payload: RuntimeApprovalActionInput,
  signal?: AbortSignal,
): Promise<RuntimeApprovalActionRead> {
  return requestV2<RuntimeApprovalActionRead>(`${runtimeApprovalPath(approvalId)}/deny`, {
    body: payload,
    method: "POST",
    signal,
  });
}

export const runtimeApi = {
  runs: {
    list: listRuntimeRuns,
    create: createRuntimeRun,
    get: getRuntimeRun,
    getArtifact: getRuntimeRunArtifact,
    getTrace: getRuntimeRunTrace,
    cancel: cancelRuntimeRun,
  },
  approvals: {
    list: listRuntimeApprovals,
    get: getRuntimeApproval,
    approve: approveRuntimeApproval,
    deny: denyRuntimeApproval,
  },
  traceEvents: {
    list: listRuntimeTraceEvents,
  },
} as const;

File: backend/app/api/runtime.py
from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, Query, status

from app.api.dependencies import get_agent_runtime_service, get_studio_query_service
from app.schemas.runtime import (
    RuntimeApprovalActionRead,
    RuntimeApprovalActionRequest,
    RuntimeApprovalListRead,
    RuntimeApprovalRead,
    RuntimeApprovalStatus,
    RuntimeArtifactRead,
    RuntimeCallerType,
    RuntimeCancelRead,
    RuntimeRunCreate,
    RuntimeRunCreated,
    RuntimeRunListRead,
    RuntimeRunRead,
    RuntimeTraceEventListRead,
    RuntimeTraceEventType,
)
from app.services.agent_runtime_service import AgentRuntimeService
from app.services.studio_query_service import StudioQueryService

router = APIRouter(prefix="/runtime", tags=["runtime"])


@router.get("/runs", response_model=RuntimeRunListRead)
def list_runtime_runs(
    service: Annotated[StudioQueryService, Depends(get_studio_query_service)],
    caller_type: Annotated[RuntimeCallerType | None, Query(alias="callerType")] = None,
    caller_id: Annotated[int | None, Query(alias="callerId")] = None,
    caller_scope_key: Annotated[str | None, Query(alias="callerScopeKey")] = None,
    caller_identity_key: Annotated[str | None, Query(alias="callerIdentityKey")] = None,
    workflow_spec_key: Annotated[str | None, Query(alias="workflowSpecKey")] = None,
) -> RuntimeRunListRead:
    return service.list_runs(
        caller_type=caller_type,
        caller_id=caller_id,
        caller_scope_key=caller_scope_key,
        caller_identity_key=caller_identity_key,
        workflow_spec_key=workflow_spec_key,
    )


@router.post("/runs", response_model=RuntimeRunCreated, status_code=status.HTTP_201_CREATED)
def create_runtime_run(
    payload: RuntimeRunCreate,
    service: Annotated[AgentRuntimeService, Depends(get_agent_runtime_service)],
) -> RuntimeRunCreated:
    return service.create_public_run(payload)


@router.get("/runs/{run_id}", response_model=RuntimeRunRead)
def get_runtime_run(
    run_id: int,
    service: Annotated[StudioQueryService, Depends(get_studio_query_service)],
) -> RuntimeRunRead:
    return service.get_run(run_id)


@router.get("/runs/{run_id}/artifacts", response_model=RuntimeArtifactRead)
def get_runtime_artifact(
    run_id: int,
    service: Annotated[StudioQueryService, Depends(get_studio_query_service)],
) -> RuntimeArtifactRead:
    return service.get_artifact(run_id)


@router.get("/runs/{run_id}/trace", response_model=RuntimeTraceEventListRead)
def get_runtime_run_trace(
    run_id: int,
    service: Annotated[StudioQueryService, Depends(get_studio_query_service)],
) -> RuntimeTraceEventListRead:
    return service.list_run_trace(run_id)


@router.post("/runs/{run_id}/cancel", response_model=RuntimeCancelRead)
def cancel_runtime_run(
    run_id: int,
    service: Annotated[AgentRuntimeService, Depends(get_agent_runtime_service)],
) -> RuntimeCancelRead:
    return service.cancel_run(run_id)


@router.get("/approvals", response_model=RuntimeApprovalListRead)
def list_runtime_approvals(
    service: Annotated[StudioQueryService, Depends(get_studio_query_service)],
    run_id: Annotated[int | None, Query(alias="runId")] = None,
    caller_type: Annotated[RuntimeCallerType | None, Query(alias="callerType")] = None,
    caller_id: Annotated[int | None, Query(alias="callerId")] = None,
    workflow_spec_key: Annotated[str | None, Query(alias="workflowSpecKey")] = None,
    capability_key: Annotated[str | None, Query(alias="capabilityKey")] = None,
    approval_status: Annotated[RuntimeApprovalStatus | None, Query(alias="status")] = None,
) -> RuntimeApprovalListRead:
    return service.list_approvals(
        run_id=run_id,
        caller_type=caller_type,
        caller_id=caller_id,
        workflow_spec_key=workflow_spec_key,
        capability_key=capability_key,
        status=approval_status,
    )


@router.get("/approvals/{approval_id}", response_model=RuntimeApprovalRead)
def get_runtime_approval(
    approval_id: int,
    service: Annotated[StudioQueryService, Depends(get_studio_query_service)],
) -> RuntimeApprovalRead:
    return service.get_approval(approval_id)


@router.get("/trace-events", response_model=RuntimeTraceEventListRead)
def list_runtime_trace_events(
    service: Annotated[StudioQueryService, Depends(get_studio_query_service)],
    run_id: Annotated[int | None, Query(alias="runId")] = None,
    caller_type: Annotated[RuntimeCallerType | None, Query(alias="callerType")] = None,
    caller_id: Annotated[int | None, Query(alias="callerId")] = None,
    workflow_spec_key: Annotated[str | None, Query(alias="workflowSpecKey")] = None,
    capability_key: Annotated[str | None, Query(alias="capabilityKey")] = None,
    event_type: Annotated[RuntimeTraceEventType | None, Query(alias="eventType")] = None,
) -> RuntimeTraceEventListRead:
    return service.list_trace_events(
        run_id=run_id,
        caller_type=caller_type,
        caller_id=caller_id,
        workflow_spec_key=workflow_spec_key,
        capability_key=capability_key,
        event_type=event_type,
    )


@router.post("/approvals/{approval_id}/approve", response_model=RuntimeApprovalActionRead)
def approve_runtime_approval(
    approval_id: int,
    payload: RuntimeApprovalActionRequest,
    service: Annotated[AgentRuntimeService, Depends(get_agent_runtime_service)],
) -> RuntimeApprovalActionRead:
    return service.approve_approval(
        approval_id,
        actor=payload.actor,
        reason=payload.reason,
    )


@router.post("/approvals/{approval_id}/deny", response_model=RuntimeApprovalActionRead)
def deny_runtime_approval(
    approval_id: int,
    payload: RuntimeApprovalActionRequest,
    service: Annotated[AgentRuntimeService, Depends(get_agent_runtime_service)],
) -> RuntimeApprovalActionRead:
    return service.deny_approval(
        approval_id,
        actor=payload.actor,
        reason=payload.reason,
    )
